### PR TITLE
CST-AST-lowering of expressions

### PR DESCRIPTION
**Pull request by JonasWanke**

### CODE REVIEW

File: packages/ast/src/lowering.candy [hunk 1]
```diff
@@ -1370,20 +1432,333 @@ fun lowerInterpolatedStringPart(
...
+  cst: CstNode<CstLambdaExpression>,
+): Maybe<AstExpression> {
+  let id = context.getExpressionId(parentId)
+
+  let valueParameters = cst.child.valueParameters
+      .map<List<AstValueParameter>>({ lowerLambdaExpressionValueParameters(context, id, it) })
+      .orElse({ List.empty<AstValueParameter>() })
+
+  let expressions = (cst.child.expressions as Iterable<CstNode<CstExpression>>)
+      .map<Maybe<AstExpression>>({ lowerExpression(context, id, it)})
```
Comment (MarcelGarus) [inline packages/ast/src/lowering.candy]:
```suggestion
      .map<Maybe<AstExpression>>({ lowerExpression(context, id, it) })
```

File: packages/ast/src/lowering.candy [hunk 2]
```diff
@@ -1370,20 +1432,333 @@ fun lowerInterpolatedStringPart(
...
+        // TODO(JonasWanke)
+      } else {
+        punctuationExpectedNext = false
+      }
+    } else {
+      if punctuationExpectedNext {
+        // TODO(JonasWanke)
+      }
+      let type = lowerExpression(context, parentId, node.cast<CstExpression>())
+      if type is Some<AstExpression> { expressions.append(type.unwrap()) }
```
Comment (MarcelGarus) [inline packages/ast/src/lowering.candy]:
```suggestion
      let expression = lowerExpression(context, parentId, node.cast<CstExpression>())
      if expression is Some<AstExpression> { expressions.append(expression.unwrap()) }
      // TODO(marcelgarus): If expression is None, add error and set isValid to false
```

File: packages/ast/src/lowering.candy [hunk 3]
```diff
@@ -1370,20 +1432,333 @@ fun lowerInterpolatedStringPart(
...
+
+  let operatorId = AstIdentifierId(id, 0)
+  context.addIdMapping(cst.child.operator.id, operatorId)
+  let operator = AstIdentifier(operatorId, cst.child.operator.child.punctuation)
+
+  let rightOperand = cst.child.rightOperand
+      .flatMap<AstExpression>({ lowerExpression(context, id, it) })
+
+  context.addIdMapping(cst.id, id)
+  Some<AstExpression>(AstBinaryExpression(id, leftOperand, operator, rightOperand))
```
Comment (MarcelGarus) [inline packages/ast/src/lowering.candy]: How does this work? If I'm not mistaken, the operands are still wrapped in a `Maybe` and the `AstBinaryExpression` should require actual expressions, shouldn't it?

File: packages/ast/src/lowering.candy [hunk 4]
```diff
@@ -1370,20 +1432,333 @@ fun lowerInterpolatedStringPart(
...
+  let receiver = lowerExpression(context, id, cst.child.receiver)
+
+  let typeArguments = cst.child.typeArguments
+      .map<List<AstTypeArgument>>({ lowerTypeArguments(context, id, it) })
+      .orElse({ List.empty<AstTypeArgument>() })
+
+  let valueArguments = lowerValueArguments(context, id, cst.child.valueArguments)
+
+  context.addIdMapping(cst.id, id)
+  Some<AstExpression>(AstCallExpression(id, receiver, typeArguments, valueArguments))
```
Comment (MarcelGarus) [inline packages/ast/src/lowering.candy]: Again, can the `receiver` and `valueArguments` be `None`?

File: packages/ast/src/lowering.candy [hunk 5]
```diff
@@ -1370,20 +1432,333 @@ fun lowerInterpolatedStringPart(
...
+        ExpectedIdentifierCompilerError(
+          context.file,
+          cst.cast<CstNavigationExpression | CstValueArgument>(),
+        ),
+      )
+      None<AstIdentifier>()
+    } else {
+      Some<AstIdentifier>(lowerIdentifier(context, id, name.first.unwrap(), 0))
+    }
+  }
```
Comment (MarcelGarus) [inline packages/ast/src/lowering.candy]:
```suggestion
  let name = cst.child.name.map<AstIdentifier>({
    it.first
      .map<AstIdentifier>({ lowerIdentifier(context, id, it, 0) })
      .orElse({
        context.addError(
          ExpectedIdentifierCompilerError(
            context.file,
            cst.cast<CstNavigationExpression | CstValueArgument>(),
          ),
        )
        None<AstIdentifier>()
      })
  })
```

**Review by MarcelGarus — CHANGES_REQUESTED** [review]

File: packages/ast/src/lowering.candy (hunk 4 above)
Comment (JonasWanke) [reply packages/ast/src/lowering.candy]: Fixed for the `receiver`. `valueArguments` is a `List<AstValueArgument>`, and hence cannot be `None`.

File: packages/ast/src/lowering.candy (hunk 2 above)
Comment (JonasWanke) [reply packages/ast/src/lowering.candy]: The lowering is now simplified and `isValid` is set appropriately. If `expression` is `None`, an error has already be generated when lowering that expression.

File: packages/ast/src/lowering.candy [hunk 6]
```diff
@@ -650,7 +656,9 @@ fun lowerBlockBody(
   let id = AstBodyId(parentId)
 
   let expressions = (cst.child.expressions as Iterable<CstNode<CstExpression>>)
-      .map<AstExpression>({ lowerExpression(context, parentId, it) })
+      .map<Maybe<AstExpression>>({ lowerExpression(context, parentId, it) })
+      .where({ it is Some<AstExpression> })
+      .map<AstExpression>({ it.unwrap() })
```
Comment (MarcelGarus) [inline packages/ast/src/lowering.candy]:
```suggestion
      .maybeMap<AstExpression>({ it })
```

File: packages/ast/src/lowering.candy [hunk 7]
```diff
@@ -1370,18 +1433,335 @@ fun lowerInterpolatedStringPart(
...
+  let id = context.getExpressionId(parentId)
+
+  let valueParameters = cst.child.valueParameters
+      .map<List<AstValueParameter>>({ lowerLambdaExpressionValueParameters(context, id, it) })
+      .orElse({ List.empty<AstValueParameter>() })
+
+  let expressions = (cst.child.expressions as Iterable<CstNode<CstExpression>>)
+      .map<Maybe<AstExpression>>({ lowerExpression(context, id, it) })
+      .where({ it is Some<AstExpression> })
+      .map<AstExpression>({ it.unwrap() })
```
Comment (MarcelGarus) [inline packages/ast/src/lowering.candy]:
```suggestion
      .maybeMap<AstExpression>({ it.unwrap() })
```

**Review by MarcelGarus — CHANGES_REQUESTED** [review]

**Review by MarcelGarus — APPROVED** [review]

**Pull request merged**